quickwit-metastore/src/metastore/single_file_metastore/mod.rs

```diff
@@ -0,0 +1,575 @@
+use crate::metastore::single_file_metastore::store_operations::{
+    delete_metadata_set, fetch_metadata_set, index_exists, put_metadata_set,
+};
+use crate::{
+    IndexMetadata, Metastore, MetastoreError, MetastoreResult, Split, SplitMetadata, SplitState,
+};
+
+/// Single file metastore implementation.
+pub struct SingleFileMetastore {
+    storage: Arc<dyn Storage>,
+    // try a rwlock
+    per_index_metastores: RwLock<HashMap<String, Arc<Mutex<MetadataSet>>>>,
+}
+
+impl SingleFileMetastore {
+    /// Creates a [`SingleFileMetastore`] for tests.
+    #[doc(hidden)]
+    pub fn for_test() -> Self {
+        use quickwit_storage::RamStorage;
+        SingleFileMetastore::new(Arc::new(RamStorage::default()))
+    }
+
+    #[cfg(test)]
+    pub fn storage(&self) -> Arc<dyn Storage> {
+        self.storage.clone()
+    }
+
+    /// Creates a [`SingleFileMetastore`] for a specified storage.
+    pub fn new(storage: Arc<dyn Storage>) -> Self {
+        Self {
+            storage,
+            per_index_metastores: Default::default(),
+        }
+    }
+
+    async fn mutate(
+        &self,
+        index_id: &str,
+        mutation: impl FnOnce(&mut MetadataSet) -> crate::MetastoreResult<bool>,
+    ) -> MetastoreResult<()> {
+        let mut metadata_set_lock = self.metadata_set_locked(index_id).await?;
+
+        let mut metadata_set = metadata_set_lock.clone();
+        let has_changed = mutation(&mut metadata_set)?;
+        if !has_changed {
+            return Ok(());
+        }
+
+        let put_result = put_metadata_set(&*self.storage, &metadata_set).await;
+        match put_result {
+            Ok(()) => {
+                *metadata_set_lock = metadata_set;
+                Ok(())
+            }
+            err @ Err(_) => {
+                // metadata_set_mutex.discard();
+                // For some of the error type here, we cannot know for sure
+                // whether the content was written or not.
+                //
+                // Just to be sure, let's discard the cache.
+                let mut per_index_metastores_wlock = self.per_index_metastores.write().await;
+
+                // At this point, we hold both locks.
+                per_index_metastores_wlock.remove(index_id);
+                metadata_set_lock.discarded = true;
+
+                err
+            }
+        }
+    }
+
+    async fn read<T, F>(&self, index_id: &str, view: F) -> MetastoreResult<T>
+    where F: FnOnce(&MetadataSet) -> MetastoreResult<T> {
+        let metadata_set_lock = self.metadata_set_locked(index_id).await?;
+        view(&*metadata_set_lock)
+    }
+
+    /// Returns a valid metadataset that is locked.
+    ///
+    /// This function guarantees that the metadataset has not been
+    /// marked as discarded.
+    async fn metadata_set_locked(
+        &self,
+        index_id: &str,
+    ) -> MetastoreResult<OwnedMutexGuard<MetadataSet>> {
+        loop {
```

Comment:
I think we could  avoid the busy loop with tokio's equivalent of `CondVar` https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html
It will be a bit more complex and less readable though.